api/src/controllers/post/administrative-penalty.js

```diff
@@ -32,53 +33,67 @@ const BusinessLogicManager = require('../../utils/business-logic-manager');
  */
-exports.createRecord = async function(args, res, next, incomingObj) {
-  // save flavour records
-  let observables = [];
-  let savedFlavourAdministrativePenalties = [];
+exports.createRecord = async function (args, res, next, incomingObj) {
+  let flavours = [];
   let flavourIds = [];
+  let observables = [];
+  // We have this in case there's error and we need to clean up.
+  let idsToDelete = [];
+
+  // Prepare flavours
+  incomingObj.AdministrativePenaltyLNG &&
+    flavours.push(this.createLNG(args, res, next, { ...incomingObj, ...incomingObj.AdministrativePenaltyLNG }));
+  incomingObj.AdministrativePenaltyNRCED &&
+    flavours.push(
+      this.createNRCED(args, res, next, { ...incomingObj, ...incomingObj.AdministrativePenaltyNRCED })
+    );
+
+  // Get flavour ids for master
+  if (flavours.length > 0) {
+    flavourIds = flavours.map(
+      flavour => flavour._id
+    );
+    idsToDelete = [...flavourIds];
+  }
 
-  try {
-    incomingObj.AdministrativePenaltyLNG &&
-      observables.push(this.createLNG(args, res, next, { ...incomingObj, ...incomingObj.AdministrativePenaltyLNG }));
-    incomingObj.AdministrativePenaltyNRCED &&
-      observables.push(
-        this.createNRCED(args, res, next, { ...incomingObj, ...incomingObj.AdministrativePenaltyNRCED })
-      );
-
-    if (observables.length > 0) {
-      savedFlavourAdministrativePenalties = await Promise.all(observables);
-
-      flavourIds = savedFlavourAdministrativePenalties.map(
-        flavourAdministrativePenalty => flavourAdministrativePenalty._id
-      );
-    }
-  } catch (e) {
-    return {
-      status: 'failure',
-      object: savedFlavourAdministrativePenalties,
-      errorMessage: e.message
-    };
+  // Prepare master
+  let masterRecord = this.createMaster(args, res, next, incomingObj, flavourIds);
+  idsToDelete.push(masterRecord._id);
+
+  // Set master back ref to flavours get ready to save
+  for (let i = 0; i < flavours.length; i++) {
+    flavours[i]._master = new ObjectId(masterRecord._id);
+    observables.push(flavours[i].save());
   }
+  observables.push(masterRecord.save());
 
-  // save administrativePenalty record
-  let savedAdministrativePenalty = null;
+  // Attempt to save everything.
 
+  let result = null;
   try {
-    savedAdministrativePenalty = await this.createMaster(args, res, next, incomingObj, flavourIds);
-
-    return {
-      status: 'success',
-      object: savedAdministrativePenalty,
-      flavours: savedFlavourAdministrativePenalties
-    };
+    result = await Promise.all(observables);
   } catch (e) {
+    // Something went wrong. Attempt to clean up
+    const db = mongodb.connection.db(process.env.MONGODB_DATABASE || 'nrpti-dev');
+    const collection = db.collection('nrpti');
+    let orArray = [];
+    for (let i = 0; i < idsToDelete.length; i++) {
+      orArray.push({ _id: new ObjectId(idsToDelete[i]) });
+    }
+    await collection.deleteMany({
+      $or: orArray
+    });
+
```

Comment:
I think we can reduce this to a common function in order to reduce the copy-paste going on here.

Reply:
ec1f4f4d0936283400ecee9070a336d5dafecd53